\documentclass[11pt]{article}
%\usepackage{fancyhdr}
\usepackage[hmargin=3cm,vmargin=3cm]{geometry}
\usepackage{dsfont}
\usepackage{bbold}
\usepackage{graphicx, float}
\usepackage{verbatim}
\usepackage{amssymb, amsmath}
%\usepackage[hmargin=3cm,vmargin=3cm]{geometry}
\usepackage{wrapfig}
\DeclareGraphicsExtensions{.pdf,.png,.jpg, .jpeg}

\begin{document}
\SweaveOpts{concordance=TRUE}

\title{Relationship between Intensity and GC Content}
\author{Subhrangshu Nandi\\
  Department of Statistics\\
  Laboratory of Molecular and Computational Genomics\\
  nandi@stat.wisc.edu}
%\date{July 15, 2013}
\maketitle
\noindent
\section{Introduction and Data description}
This is the first report of the work on the relationship between {\emph{flouroscence\_intensity}} and sequence composition or {\emph{GC-content}}. This is based on the data produced by Steve which includes $5,000$ groups. A {\emph{group}} is a channel through which the molecules are stretched and photographed. A {\emph{group}} can be associated with an average of $50$ molecules. This dataset has $3,871,094$ observations, with each observation being that of a fragment that has been aligned to a reference, using the optical mapping aligner. Each observation has the {\emph{flouroscence\_intensity}} of the fragment, the {\emph{GC-content}} of the corresponding location on the reference genome, and other variables such as:
\begin{itemize}
\item
{\emph{moleculeID}}: Which molecule the fragment is a part of
\item
{\emph{groupID}}: Which channel on the surface the molecule was photographed on
\item
{\emph{alignedChr}}: Aligned chromosome (1, 2, ..., 23, X, Y)
\item
{\emph{alignedFragIndex}}: Location index of the fragment for each chromosome. This variable along with {\emph{alignedChr}} uniquely identifies a genomic location. There could be multiple molecules aligned to the same gemomic location and one of the goals of the study is to use these observations to eliminate possible sources of error.
\item
{\emph{numFrags}}: Number of fragments the molecule was divided into
\item
{\emph{numPixels}}: A measure of the length of the molecule, higher the value, longer the molecule
\end{itemize}

\noindent
This dataset has $231,120$ molecules, observed in $4,821$ groups. 

\section{Step 1: Analyze fragement aligned to same genome location}
The first goal is to identify identical fragments (i.e., those that have been aligned to the same genome location) from multiple molecules. In order to do this, the orginal dataset is aggregated on a fragment level and the mean and standard deviations of the {\emph{flouroscence\_intensities}} of the fragments are estimated. A snapshot of this aggregated dataset is shown below:

<<echo=FALSE>>=
rm(list=ls(all.names=TRUE))
rm(list=objects(all.names=TRUE))
#dev.off()

########################################################################
## This script is for conducting further analysis on intensity and gc ##
## content. This will use data produced by RScript03.This script will ##
## incorporate fragment level effect first, as per discussion during  ##
## the meeting with M. Newton and D. Schwartz on 07/15                ##
########################################################################

########################################################################
## Initialize Header file and function library                        ##
########################################################################
FilePath <- '~/Project_GC_Content/RScripts_GC_Content/'
DataPath <- '~/Project_GC_Content/'
DataPath.Steve <- '/aspen/steveg/human_nMaps/GC_content/'
Filename.Header <- paste('~/RScripts/HeaderFile_lmcg.R', sep='')
source(Filename.Header)
source(paste(FilePath, 'fn_Library_GC_Content.R', sep=''))
########################################################################

########################################################################
## Data Input                                                         ##
########################################################################
filename='FragmentData_5000.RData'
Filename <- paste(FilePath, filename, sep='')
load(Filename)
FragmentData$fractionGC_sd <- NULL
names(FragmentData)[names(FragmentData)=='fractionGC_mean'] <- 'fractionGC'
rownames(FragmentData) <- NULL
head(FragmentData)
@
So, from the first row of this table, we notice that the coverage of aligned fragment index number 8576 of chromosome 16 is 86. This particular fragment has a GC-content of $48.51\%$ and the mean and standard deviations of the fluoroscence intensities are $36,945.85$ and $4,719.46$, respectively. There are a total of $308,545$ fragments. However, the coverage is not as high as 86 for most of them. Below is a summary of the number of molecules and a histogram plot (Fig 1) of the coverage of these fragments:
<<echo=FALSE>>=
print('Summary of Coverage of fragments: ')
print(summary(FragmentData$numMolecules))
@
\begin{figure}[th]
\centering 
<<echo=FALSE, fig=TRUE>>=
hist(FragmentData$numMolecules, breaks=40, main='Histogram of Coverage of fragments', xlab='Number of molecules')
@
% \caption{Some exploratory plots of the Data} % title of Table
% \label{fig:Fig1}
\end{figure}

When the variable $intensity\_mean$ is regressed with $fractionGC\_mean$ (or $GC-content$), the coefficient is quite significant (with p-value $<e^{-16}$, but the overall $R^2$ is only $2.5\%$. Similary, $intensity\_sd$ also has a significant relationship with $GC-content$, but the model $R^2$ is quite poor. There is evidence of a positive relationship between fluoroscence intensity and GC-content, however, more in-depth analysis is required. 

\section{Step 2: Modeling intensity and GC-content}
Prof. Schwartz recommended that we analyze only those fragments that have at least 40 coverage. Hence, the following analysis is conducted on the original dataset trimmmed down to only those fragments that have at least 40 molecules aligned to the same genomic location. This trimmed dataset has $35,040$ molecules and $4,513$ groups. The following models were fit:
\begin{enumerate}
\item
$Model1:\\ Y = log(intensity) \\ X = gc\_content$
\item
$Model2:\\ Y = log(intensity) \\ X = gc\_content + numFrags$ \\
numFrags is a measure of the number of fragments the molecule was divided into.
\item
$Model3:\\ Y = log(intensity) \\ X = gc\_content + numFrags + as.factor(groupID)$ \\
This model is to study how much of variability the channels on the surface contribute to that of the observed fluoroscence intensity.
\item
$Model4:\\ Y = log(intensity) \\ X = gc\_content + numFrags + as.factor(groupID) + numPixels$ \\
numPixels is a measure of the length (or size) of the molecule. Higher the number of pixels, longer the molecule. 
\end{enumerate}

Below are the R-outputs of the above mentioned models, and their anovas.\\
\noindent
{\bf{\underline{Summary of Model 1}}}
<<echo=FALSE>>=
Filename.Model1 <- paste(FilePath, 'Model1.RData', sep='')
Filename.Model2 <- paste(FilePath, 'Model2.RData', sep='')
Filename.Model4 <- paste(FilePath, 'Model4.RData', sep='')
load(Filename.Model1)
load(Filename.Model2)
summary(Model1)
@
\noindent
{\bf{\underline{Summary of Model 2:}}}
<<echo=FALSE>>=
summary(Model2)
@
\noindent
{\bf{\underline{Anova between Model1 and Model2}}}
<<echo=FALSE>>=
anova(Model1, Model2)
@

\noindent
{\bf{\underline{Summary of Model 3}}}
<<echo=FALSE>>=
Summary.Model3 <- paste(FilePath, 'Model3_Summary.RData', sep='')
load(Summary.Model3)
Anova.Model3 <- paste(FilePath, 'Model3_Anova.RData', sep='')
load(Anova.Model3)
print(Summary3$coefficients[1:4,])
print('R-Squared of Model 3:')
print(Summary3$r.squared)
@

\noindent
{\bf{\underline{Anova between Model2 and Model3}}}
<<echo=FALSE>>=
print(Anova3)
@

\noindent
{\bf{\underline{Summary of Model 4}}}
<<echo=FALSE>>=
Summary.Model4 <- paste(FilePath, 'Model4_Summary.RData', sep='')
load(Summary.Model4)
Anova.Model4 <- paste(FilePath, 'Model4_Anova.RData', sep='')
load(Anova.Model4)
print(Summary4$coefficients[1:4,])
print('R-Squared of Model 4:')
print(Summary4$r.squared)
@

\noindent
{\bf{\underline{Anova between Model3 and Model4}}}
<<echo=FALSE>>=
print(Anova4)
@

\noindent
{\bf{\underline{Conclusion}}}: Introducing the groups as factors seem to explain $60\%$ of the variability. Since there are more than $35,000$ molecules, we can only check the molecule level effect for a few 100 molecules at a time. Given that so much variability in the response is a result of experimental artifacts, do you think there is any value proceeding with this project? Interestingly, the relationship between fluoroscence intensity and gc\_content remains significant even after introducing the group level effects.

\end{document}
